pkg/component/etcd/statefulset/statefulset.go

```diff
@@ -200,6 +214,107 @@ func (c *component) WaitCleanup(ctx context.Context) error {
+	waitPodsTaskID := g.Add(flow.Task{
+		Name: waitPodsOpName,
+		Fn: func(ctx context.Context) error {
+			return c.waitUntilPodsUpdatedAndReady(ctx, sts, defaultInterval, defaultTimeout*2)
+		},
+		Dependencies: flow.NewTaskIDs(patchLabelsTaskID),
+	})
+	c.logger.Info("adding task to pre-deploy flow", "name", waitPodsOpName, "ID", waitPodsTaskID)
+
+	// If the selector is not as expected, then delete the StatefulSet.
+	deleteStsOpName := "(delete-sts-with-orphans): Deleting StatefulSet by orphaning pods"
+	deleteStsTaskID := g.Add(flow.Task{
+		Name: deleteStsOpName,
+		Fn: func(ctx context.Context) error {
+			return c.deleteWithOrphanCascade(ctx, sts)
+		},
+		Dependencies: flow.NewTaskIDs(waitPodsTaskID),
+	})
+	c.logger.Info("adding task to pre-deploy flow", "name", deleteStsOpName, "ID", deleteStsTaskID)
+}
+
+func (c *component) patchPodTemplateLabels(ctx context.Context, sts *appsv1.StatefulSet) error {
+	if !utils.ContainsAllDesiredLabels(sts.Spec.Template.Labels, c.values.PodLabels) {
+		c.logger.Info("Patching StatefulSet pod template labels", "namespace", c.values.Namespace, "name", c.values.Name, "podTemplateLabels", utils.MergeStringMaps(c.values.PodLabels, c.values.AdditionalPodLabels))
+		patch := client.MergeFrom(sts.DeepCopy())
+		sts.Spec.Template.Labels = utils.MergeStringMaps(c.values.PodLabels, c.values.AdditionalPodLabels)
+		return c.client.Patch(ctx, sts, patch)
+	}
+	return nil
+}
+
+func (c *component) waitUntilPodsUpdatedAndReady(ctx context.Context, sts *appsv1.StatefulSet, interval, timeout time.Duration) error {
+	return gardenerretry.UntilTimeout(ctx, interval, timeout, func(ctx context.Context) (bool, error) {
+		c.logger.Info("Waiting for StatefulSet pods to be updated", "namespace", c.values.Namespace, "name", c.values.Name)
+		if err := c.client.Get(ctx, client.ObjectKeyFromObject(sts), sts); err != nil {
+			return gardenerretry.MinorError(fmt.Errorf("failed to fetch sts %s: %v", sts.Name, err))
+		}
+
+		if sts.Status.ObservedGeneration < sts.Generation {
+			return gardenerretry.MinorError(fmt.Errorf("StatefulSet spec change has not yet been picked up for reconciliation by statefulset-controller"))
+		}
+		if sts.Status.UpdateRevision != sts.Status.CurrentRevision {
+			return gardenerretry.MinorError(fmt.Errorf("StatefulSet update revision %s does not match current revision %s", sts.Status.UpdateRevision, sts.Status.CurrentRevision))
+		}
+		if sts.Status.UpdatedReplicas < *sts.Spec.Replicas {
+			return gardenerretry.MinorError(fmt.Errorf("only %d out of %d replicas are updated", sts.Status.UpdatedReplicas, *sts.Spec.Replicas))
+		}
+		if sts.Status.ReadyReplicas < *sts.Spec.Replicas {
+			return gardenerretry.MinorError(fmt.Errorf("only %d out of %d replicas are ready", sts.Status.ReadyReplicas, sts.Spec.Replicas))
+		}
+		return gardenerretry.Ok()
+	})
+}
+
+func (c *component) deleteWithOrphanCascade(ctx context.Context, sts *appsv1.StatefulSet) error {
```

Comment:
Can it be documented somewhere that once the PreDeployFlow completes and the reconciliation of the Etcd starts, the newly recreated statefulset will adopt that pods that were previously orphaned through the cascade orphan delete?